src/merkle/store.rs

```diff
@@ -0,0 +1,375 @@
+            );
+        }
+        MerkleStore { nodes }
+    }
+
+    /// Creates the merkle tree from `leaves` and adds all of their elements into the store.
+    pub fn add_merkle_tree(&mut self, leaves: Vec<Word>) -> Result<Word, MerkleError> {
+        let layers = leaves.len().ilog2();
+        let tree = MerkleTree::new(leaves)?;
+
+        let mut depth = 0;
+        let mut parent_offset = 1;
+        let mut child_offset = 2;
+        while depth < layers {
+            let layer_size = 1usize << depth;
+            for _ in 0..layer_size {
+                // merkle tree is using level form representation, so left and right siblings are
+                // next to each other
+                let left = tree.nodes[child_offset];
+                let right = tree.nodes[child_offset + 1];
+                self.nodes.insert(
+                    tree.nodes[parent_offset].into(),
+                    Node {
+                        left: left.into(),
+                        right: right.into(),
+                    },
+                );
+                parent_offset += 1;
+                child_offset += 2;
+            }
+            depth += 1;
+        }
+
+        Ok(tree.nodes[1])
+    }
+
+    pub fn add_sparse_merkle_tree<R, I>(&mut self, entries: R) -> Result<Word, MerkleError>
+    where
+        R: IntoIterator<IntoIter = I>,
+        I: Iterator<Item = (u64, Word)> + ExactSizeIterator,
+    {
+        let smt = SimpleSmt::new(SimpleSmt::MAX_DEPTH)?.with_leaves(entries)?;
+        for branch in smt.store.branches.values() {
+            let parent = Rpo256::merge(&[branch.left, branch.right]);
+            self.nodes.insert(
+                parent.into(),
+                Node {
+                    left: branch.left.into(),
+                    right: branch.right.into(),
+                },
+            );
+        }
+
+        Ok(smt.root())
+    }
+
+    pub fn add_merkle_paths<I>(&mut self, paths: I) -> Result<Word, MerkleError>
+    where
+        I: IntoIterator<Item = (u64, Word, MerklePath)>,
+    {
+        let node: Word = Word::default();
+
+        for (index_value, mut node, path) in paths {
+            let mut index = NodeIndex::new(self.nodes.len() as u8, index_value);
+
+            for sibling in path {
+                let (left, right) = match index.is_value_odd() {
+                    true => (sibling, node),
+                    false => (node, sibling),
+                };
+                let parent = Rpo256::merge(&[left.into(), right.into()]).into();
+                self.nodes.insert(
+                    parent,
+                    Node {
+                        left: left.into(),
+                        right: right.into(),
+                    },
+                );
+
+                index.move_up();
+                node = parent.word;
+            }
+        }
+
+        // Returns the parent of the last paths (assumes all paths have the same parent) or empty
+        Ok(node)
```

Comment:
I think we should make sure that all paths in fact resolve to the same root.

Also, maybe it is worth creating a constructor for adding a single Merkle path?

Reply:
> Also, maybe it is worth creating a constructor for adding a single Merkle path?

done

Reply:
> I think we should make sure that all paths in fact resolve to the same root.

Added. This meant the iterator had to be collected into a vector to compute the roots prior to the comparison.

Reply:
This is my rationale: I tend to write code which validates the data first, and then performs changes. This usually means it is possible to write better error messages (in this case report all the roots that differ instead of the first two), IMO it makes it easier to write correct code since the state changes are applied after all the invariants are validated, so no partial state changes (in this case it would be inserting some but not all of the paths), and it gives a nice structure where the code starts with validation using short-circuit followed by the core logic without mixing them together.

Let me know if I should change.

Reply:
I think it is fine for now. I'd probably add a `TODO` to revisit later and optimize.